engine/src/main/java/io/zeebe/engine/processing/streamprocessor/ReProcessingStateMachine.java

```diff
@@ -342,4 +370,54 @@ private void onRecovered(final long lastProcessedPosition) {
     recoveryFuture.complete(lastProcessedPosition);
     failedEventPositions.clear();
   }
+
+  private void verifyRecordMatchesToReprocessing(final TypedRecord<?> currentEvent) {
+
+    if (currentEvent.getSourceRecordPosition() < 0
+        || currentEvent.getSourceRecordPosition() <= snapshotPosition) {
+      // ignore commands (i.e. no source currentEvent position) and records that are not produced by
+      // the reprocessing (i.e. the source currentEvent is already compacted)
+      return;
+    }
+
+    // if a record is not written to the log stream then the state could be corrupted
+    reprocessingStreamWriter.getRecords().stream()
+        .filter(record -> record.getSourceRecordPosition() < currentEvent.getSourceRecordPosition())
+        .findFirst()
+        .ifPresent(
+            missingRecordOnLogStream -> {
+              throw new InconsistentReprocessingException(
+                  "Records were created on reprocessing but not written on the log stream.",
+                  typedEvent,
+                  missingRecordOnLogStream);
+            });
+
+    // If the record was not written on reprocessing then the next record may have a different key,
+    // or the state is corrupted. But since the source record position can be wrong (#5420), we can
+    // not fail the reprocessing at the moment.
+
+    reprocessingStreamWriter.getRecords().stream()
+        .filter(
+            record -> record.getSourceRecordPosition() == currentEvent.getSourceRecordPosition())
+        .findFirst()
+        .ifPresent(
```

Comment:
If there was no such record written on reprocessing then we can't apply any check :sweat_smile: 

Because of #5420, we can not fail and assume that something went wrong.